File: day_02/main.py
def is_safe(nums):
    inc = nums[1] - nums[0]
    for n1, n2 in zip(nums, nums[1:]):
        diff = n2 - n1
        if abs(diff) < 1 or abs(diff) > 3 or (inc * diff) <= 0:
            return False
    return True


with open("input.txt") as f:
    total = 0
    for line in f:
        nums = [int(x) for x in line.split()]
        if is_safe(nums) or any(
            is_safe(nums[:i] + nums[i + 1 :]) for i in range(len(nums))
        ):
            total += 1
    print(total)  # 318

File: day_02/main.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <ranges>
#include <sstream>
#include <vector>

using namespace std;

int main() {
  fstream in{"input.txt"};
  if (!in) return 1;

  int total = 0;
  string line;
  while (getline(in, line)) {
    istringstream iss(line);
    vector<int> sequence{istream_iterator<int>(iss), istream_iterator<int>()};
    int incr = sequence[1] - sequence[0];
    bool safe = true;
    for (auto [n1, n2] : views::zip(sequence, sequence | views::drop(1))) {
      cout << "n1 " << n1 << " n2 " << n2 << endl;
      int diff = n2 - n1;
      if (abs(diff) < 1 || abs(diff) > 3 or (incr * diff) <= 0) {
        safe = false;
        break;
      }
    }
    cout << "safe: " << safe << endl;
    if (safe) total++;
  }
  cout << total << endl;
  return 0;
}
